documents/design/oavb_ipc.asciidoc

```diff
@@ -0,0 +1,83 @@
+= Interprocess Communication Recommendations for AVB/TSN
+Andrew Elder
+v0.1
+
+== Introduction
+
+The AVB/TSN software stack is comprised of a number of modules that implement protocol stacks specified by IEEE standards. Each of these protocols implement an Ethernet network facing interface and an application facing interface. The required application facing interface is described in some detail in http://avnu.org/wp-content/uploads/2014/05/AVnu_SWAPIs_v1.0.pdf[AVB Software Interfaces and Endpoint Architecture Guidelines]. Since a module running a particular protocol may be a standalone execution unit (for example a Linux daemon), an interprocess communication method is required to interface to the execution unit.
+
+For portability reasons it is recommended to separate the implementation of the interprocess communication method as much as possible from the logic of the protocol implementation. This applies to both an application that is interfacing to an AVB/TSN module and the implementation of the AVB/TSN module itself.
+
+For the purposes of this document the following straw-man application interfaces are described:
+
+.Straw-man Interfaces
+[source,c/c++]
+----
+
+void AVBTSN_SetParam(uint32_t param);
+
+----
+
+where the AVBTSN portion of the function name would be replaced by the name of the IEEE AVB/TSN protocol module that is being targeted.
```

Comment:
If you're going to do multiple parameters and a synchronous interface, a reasonable error mechanism is to always pass output parameters via 'out' argument pointers and use the function return value exclusively as a status code. For more detailed errors, provide an additional API function to retrieve module-specific error data via out parameters.

For an asynchronous interface, you would provide both a success callback function and an error callback function when making an API call. The nice thing about callbacks is that they can pass their parameters by value instead of requiring out-parameter pointers. The not-so-nice thing is how you have to chop your functions into several pieces and put common variables in a structure that's not stack-allocated.

I'd try to avoid specifying anything about allocation so that embedded devices can use static rather than heap allocation for context/state structures.